### PR TITLE
Fix snapshot getting stuck in INIT state

### LINKED ISSUE #27180: Snapshot process can get stuck in INIT state
If the master disconnects from the cluster after initiating snapshot, but just before the snapshot switches from INIT to STARTED state, the snapshot can get indefinitely stuck in the INIT state. This error is specific to v5.x+ and was triggered by [keeping the master node that stepped down in the node list](https://github.com/elastic/elasticsearch/pull/22049), the cleanup logic in snapshot/restore assumed that if master steps down it is always removed from the the node list. We need to change the cleanup logic to be triggered even if no nodes left the cluster.

### PR DESCRIPTION
**Pull request by imotov**
If the master disconnects from the cluster after initiating snapshot, but just before the snapshot switches from INIT to STARTED state, the snapshot can get indefinitely stuck in the INIT state. This error is specific to v5.x+ and was triggered by keeping the master node that stepped down in the node list, the cleanup logic in snapshot/restore assumed that if master steps down it is always removed from the the node list. This commit changes the logic to trigger cleanup even if no nodes left the cluster.
    
Closes #27180

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -1014,6 +1028,11 @@ public void onFailure(String source, Exception e) {
             }
 
             @Override
+            public void onNoLongerMaster(String source) {
+                listener.onNoLongerMaster(source);
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: can listener be null here? It is marked as Nullable above?

File: core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                }
+            }
+        });
+
+        logger.info("--> starting snapshot");
+        try {
+            client().admin().cluster().prepareCreateSnapshot("test-repo", "test-snap-2").setWaitForCompletion(false)
+                .setMasterNodeTimeout("5s").setIndices(idxName).get();
+        } catch (Exception ex) {
+            logger.info("--> got exception from hanged master", ex);
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java]: should this be a timeout exception. can you assert on that?

File: core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        createRandomIndex(idxName);
+
+        logger.info("-->  creating repository");
+        assertAcked(client().admin().cluster().preparePutRepository("test-repo")
+            .setType("fs").setSettings(Settings.builder()
+                .put("location", randomRepoPath())
+                .put("compress", randomBoolean())
+                .put("chunk_size", randomIntBetween(100, 1000), ByteSizeUnit.BYTES)));
+
+        // Writing incompatible snapshot can cause this test to fail due to a race condition in repo initialization
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java]: I don't understand this comment. What's the issue with repo initialization here? When the disruption triggers, then there is no more repo writing done by the old master node AFAICS?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Overall change LGTM. I've left one question and two minor comments.

File: core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (imotov) [reply core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Yes! Good catch! Fixing it.

File: core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java (hunk 3 above)
Comment (imotov) [reply core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java]: There is a race condition in writing a list of incompatible snapshot in getRepositoryData method that occurs on empty repositories. This method is called in the START phase on the former master and during clean up on the new master if this file doesn't exist in the repo, which happens in the repo. It shouldn't cause any issues in the real life, but it makes test to fail occasionally due to asserts. We will definitely need to address it at some point of time, but I don't think we should do it as part of this PR.

File: core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java (hunk 2 above)
Comment (imotov) [reply core/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java]: Good point.

**Review by ywelsch — APPROVED** [review]
LGTM. Let's have this run for a bit on CI on master before backporting.

**Pull request merged**